### PR TITLE
Remove Page/HeadPage/TailPage) inheritance

### PR DESCRIPTION
**Pull request by colinsurprenant**
This is work needed for the batch maximization #8674. 

To be able to continue reading from a page upon beheading (moving a head page into a tail page) which happen in a write operation, a *stable* Page object is required which can be used for read operations regardless if it is a head page or transformed into a tail page. 

This PR changes the idea of distinct HeadPage and TailPage objects to a single Page object which is simply `writable` or not which really only affects the ability to perform a `write` operation at all and change the `checkpoint` strategy.

All tests are passing, this refactor surfaced a concurrency issue and extra locking has been added to public Queue methods. Otherwise I minimized any other non-related code change.

Changes explanations:
- Basically `HeadPage` and `TailPage` code has been moved into `Page`,  the only real difference was with the `write()` method and `checkpoint()` which are now guarded with a new `writable` attribute.
- The `HeadPage` and `TailPage` constructors have been moved into a new `PageFactory`.
- Added locking around remaining `Queue` public methods
- Move the Page capacity check inside the `write()` method locked section

### CODE REVIEW

File: logstash-core/src/main/java/org/logstash/ackedqueue/Page.java [hunk 1]
```diff
@@ -2,27 +2,28 @@
...
-import java.util.ArrayList;
 import java.util.BitSet;
-import java.util.List;
+
+import org.logstash.ackedqueue.io.CheckpointIO;
 import org.logstash.ackedqueue.io.LongVector;
 import org.logstash.ackedqueue.io.PageIO;
 
-public abstract class Page implements Closeable {
+public class Page implements Closeable {
```
Comment (original-brownbear) [inline logstash-core/src/main/java/org/logstash/ackedqueue/Page.java]: @colinsurprenant I guess we can even make this one `final` now right?

File: logstash-core/src/main/java/org/logstash/ackedqueue/Page.java [hunk 2]
```diff
@@ -32,30 +33,52 @@ public Page(int pageNum, Queue queue, long minSeqNum, int elementCount, long fir
...
      */
     public SequencedList<byte[]> read(int limit) throws IOException {
-
         // first make sure this page is activated, activating previously activated is harmless
         this.pageIO.activate();
 
         SequencedList<byte[]> serialized = this.pageIO.read(this.firstUnreadSeqNum, limit);
         assert serialized.getSeqNums().get(0) == this.firstUnreadSeqNum :
-            String.format("firstUnreadSeqNum=%d != first result seqNum=%d", this.firstUnreadSeqNum, serialized.getSeqNums().get(0));
+                String.format("firstUnreadSeqNum=%d != first result seqNum=%d", this.firstUnreadSeqNum, serialized.getSeqNums().get(0));
```
Comment (original-brownbear) [inline logstash-core/src/main/java/org/logstash/ackedqueue/Page.java]: @colinsurprenant nit: this indent looks broken now?

File: logstash-core/src/main/java/org/logstash/ackedqueue/Page.java [hunk 3]
```diff
@@ -32,30 +33,52 @@ public Page(int pageNum, Queue queue, long minSeqNum, int elementCount, long fir
...
 
         this.firstUnreadSeqNum += serialized.getElements().size();
 
         return serialized;
     }
 
+    public void write(byte[] bytes, long seqNum, int checkpointMaxWrites) throws IOException {
+        if (! this.writable) {
+            throw new IOException(String.format("page=%d is not writable", this.pageNum));
```
Comment (original-brownbear) [inline logstash-core/src/main/java/org/logstash/ackedqueue/Page.java]: @colinsurprenant This isn't an `IOException` is it? We're throwing this as a result of a logic error (trying to write to a tail page) as far as I can tell. If so we must throw a `RuntimeException` like `IllegalStateException`.

File: logstash-core/src/main/java/org/logstash/ackedqueue/PageFactory.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+    /**
+     * create a new head page from an existing {@link Checkpoint} and open page.{@literal {pageNum}} empty valid data file
+     *
+     * @param checkpoint existing head page {@link Checkpoint}
+     * @param queue the {@link Queue} instance
+     * @param pageIO the {@link PageIO} delegate
+     * @return {@link Page} the new head page
+     */
+    public static Page newHeadPage(Checkpoint checkpoint, Queue queue, PageIO pageIO) {
```
Comment (original-brownbear) [inline logstash-core/src/main/java/org/logstash/ackedqueue/PageFactory.java]: @colinsurprenant the whole body of this method should probably be wrapped in a `try { ... } catch (Exception ex) {p.close; throw ex;}` Otherwise any exception thrown in the body will leave a `Page` behind that goes out of scope without a `close`.

File: logstash-core/src/main/java/org/logstash/ackedqueue/PageFactory.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+    /**
+     * create a new tail page for an exiting Checkpoint and data file
+     *
+     * @param checkpoint existing tail page {@link Checkpoint}
+     * @param queue the {@link Queue} instance
+     * @param pageIO the {@link PageIO} delegate
+     * @return {@link Page} the new tail page
+     */
+    public static Page newTailPage(Checkpoint checkpoint, Queue queue, PageIO pageIO) {
```
Comment (original-brownbear) [inline logstash-core/src/main/java/org/logstash/ackedqueue/PageFactory.java]: @colinsurprenant the whole body of this method should probably be wrapped in a `try { ... } catch (Exception ex) {p.close; throw ex;}` Otherwise any exception thrown in the body will leave a `Page` behind that goes out of scope without a `close`.

Comment (original-brownbear) [reply logstash-core/src/main/java/org/logstash/ackedqueue/PageFactory.java]: @colinsurprenant also, wouldn't this be better called `openTailPage` since we're not actually creating a file? :)

File: logstash-core/src/main/java/org/logstash/ackedqueue/PageFactory.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    /**
+     * create a new head page from an existing {@link Checkpoint} and open page.{@literal {pageNum}} empty valid data file
+     *
+     * @param checkpoint existing head page {@link Checkpoint}
+     * @param queue the {@link Queue} instance
+     * @param pageIO the {@link PageIO} delegate
+     * @return {@link Page} the new head page
+     */
+    public static Page newHeadPage(Checkpoint checkpoint, Queue queue, PageIO pageIO) {
+        Page p = new Page(
```
Comment (original-brownbear) [inline logstash-core/src/main/java/org/logstash/ackedqueue/PageFactory.java]: @colinsurprenant can't we simply overload the constructor of `Page` and save the duplication between this method and `newTailPage`?

**Comment by original-brownbear** [comment]
@colinsurprenant looks good for the most part, **much** cleaner and easier to follow with just `Page` :)
Just a few small comments around `Exception`s.

File: logstash-core/src/main/java/org/logstash/ackedqueue/PageFactory.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
+package org.logstash.ackedqueue;
+
+import org.logstash.ackedqueue.io.PageIO;
+
+import java.util.BitSet;
+
+public class PageFactory {
```
Comment (original-brownbear) [inline logstash-core/src/main/java/org/logstash/ackedqueue/PageFactory.java]: @colinsurprenant slightly unrelated topic to this PR, but since we're going to expose our Java codebase eventually (hopefully soon :D), do we need to have this new class `public`? I think package private suffices here since only `Queue` uses it right?

**Comment by colinsurprenant** [comment]
@original-brownbear thanks for the review! good comments, will address shortly!

File: logstash-core/src/main/java/org/logstash/ackedqueue/Page.java (hunk 1 above)
Comment (colinsurprenant) [reply logstash-core/src/main/java/org/logstash/ackedqueue/Page.java]: +1

File: logstash-core/src/main/java/org/logstash/ackedqueue/Page.java (hunk 2 above)
Comment (colinsurprenant) [reply logstash-core/src/main/java/org/logstash/ackedqueue/Page.java]: +1

File: logstash-core/src/main/java/org/logstash/ackedqueue/PageFactory.java (hunk 4 above)
Comment (colinsurprenant) [reply logstash-core/src/main/java/org/logstash/ackedqueue/PageFactory.java]: good idea.

File: logstash-core/src/main/java/org/logstash/ackedqueue/PageFactory.java (hunk 5 above)
Comment (colinsurprenant) [reply logstash-core/src/main/java/org/logstash/ackedqueue/PageFactory.java]: ah ... good question, new in this context is about the object instantiation, like a constructor, not about the page file creation.

File: logstash-core/src/main/java/org/logstash/ackedqueue/PageFactory.java (hunk 6 above)
Comment (colinsurprenant) [reply logstash-core/src/main/java/org/logstash/ackedqueue/PageFactory.java]: we could be I'd prefer to keep to so that it is obvious in the code that we are creating a new head page instead of creating a page with a `true` value for the `writable` parameter.

**Comment by colinsurprenant** [comment]
@original-brownbear  pushed review fixes LMKWYT.

**Comment by original-brownbear** [comment]
@colinsurprenant LGTM (don't forget to squash :D)

**Comment by colinsurprenant** [comment]
Thanks @original-brownbear ! 

merged into master and 6.x
